File: GameServer/src/gameserver/skillengine/model/SkillTemplate.java
/*
 * This file is part of Aion X EMU <aionxemu>.
 *
 *  This is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software.  If not, see <http://www.gnu.org/licenses/>.
 */
package gameserver.skillengine.model;

import gameserver.skillengine.action.Actions;
import gameserver.skillengine.condition.Conditions;
import gameserver.skillengine.effect.EffectTemplate;
import gameserver.skillengine.effect.Effects;
import gameserver.skillengine.properties.Properties;
import gameserver.skillengine.properties.Property;
import gameserver.skillengine.properties.TargetRangeProperty;
import gameserver.skillengine.properties.TargetRelationProperty;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlType;


/**
 * @author ATracer
 */
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "skillTemplate", propOrder = {
        "initproperties",
        "startconditions",
        "setproperties",
        "useconditions",
        "effects",
        "actions"
})
public class SkillTemplate {
    protected Properties initproperties;
    protected Conditions startconditions;
    protected Properties setproperties;
    protected Conditions useconditions;
    protected Effects effects;
    protected Actions actions;


    @XmlAttribute(name = "skill_id", required = true)
    protected int skillId;
    @XmlAttribute(name = "name", required = true)
    protected String name;
    @XmlAttribute(name = "nameId", required = true)
    protected int nameId;
    @XmlAttribute(name = "stack")
    protected String stack = "NONE";
    @XmlAttribute(name = "lvl")
    protected int lvl;
    @XmlAttribute(name = "skilltype", required = true)
    protected SkillType type;
    @XmlAttribute(name = "skillsubtype", required = true)
    protected SkillSubType subType;
    @XmlAttribute(name = "tslot")
    protected SkillTargetSlot targetSlot;
    @XmlAttribute(name = "tslot_level")
    protected int targetSlotLevel;
    @XmlAttribute(name = "activation", required = true)
    protected ActivationAttribute activationAttribute;
    @XmlAttribute(name = "duration", required = true)
    protected int duration;
    @XmlAttribute(name = "cooldown")
    protected int cooldown;
    @XmlAttribute(name = "penalty_skill_id")
    protected int penaltySkillId;
    @XmlAttribute(name = "pvp_damage")
    protected int pvpDamage;
    @XmlAttribute(name = "pvp_duration")
    protected int pvpDuration;
    @XmlAttribute(name = "chain_skill_prob")
    protected int chainSkillProb;
    @XmlAttribute(name = "cancel_rate")
    protected int cancelRate;
    @XmlAttribute(name = "stance")
    protected boolean stance;
    @XmlAttribute(name = "skillset_exception")
//<delay_id> in client_skills??
    protected int skillSetException;

    /**
     * @return the initProperties
     */
    public Properties getInitproperties() {
        return initproperties;
    }

    /**
     * @return the setProperties
     */
    public Properties getSetproperties() {
        return setproperties;
    }

    /**
     * Gets the value of the startconditions property.
     *
     * @return possible object is
     *         {@link Conditions }
     */
    public Conditions getStartconditions() {
        return startconditions;
    }

    /**
     * Gets the value of the useconditions property.
     *
     * @return possible object is
     *         {@link Conditions }
     */
    public Conditions getUseconditions() {
        return useconditions;
    }

    /**
     * Gets the value of the effects property.
     *
     * @return possible object is
     *         {@link Effects }
     */
    public Effects getEffects() {
        return effects;
    }

    /**
     * Gets the value of the actions property.
     *
     * @return possible object is
     *         {@link Actions }
     */
    public Actions getActions() {
        return actions;
    }

    /**
     * Gets the value of the skillId property.
     */
    public int getSkillId() {
        return skillId;
    }

    /**
     * Gets the value of the name property.
     *
     * @return possible object is
     *         {@link String }
     */
    public String getName() {
        return name;
    }

    /**
     * @return the nameId
     */
    public int getNameId() {
        return nameId;
    }

    /**
     * @return the stack
     */
    public String getStack() {
        return stack;
    }

    /**
     * @return the lvl
     */
    public int getLvl() {
        return lvl;
    }

    /**
     * Gets the value of the type property.
     *
     * @return possible object is
     *         {@link SkillType }
     */
    public SkillType getType() {
        return type;
    }

    /**
     * @return the subType
     */
    public SkillSubType getSubType() {
        return subType;
    }

    /**
     * @return the targetSlot
     */
    public SkillTargetSlot getTargetSlot() {
        return targetSlot;
    }

    /**
     * @return the targetSlot Level
     */
    public int getTargetSlotLevel() {
        return targetSlotLevel;
    }

    /**
     * @return the duration
     */
    public int getDuration() {
        return duration;
    }

    /**
     * @return the activationAttribute
     */
    public ActivationAttribute getActivationAttribute() {
        return activationAttribute;
    }

    public boolean isPassive() {
        return activationAttribute == ActivationAttribute.PASSIVE;
    }

    public boolean isToggle() {
        return activationAttribute == ActivationAttribute.TOGGLE;
    }

    public boolean isProvoked() {
        return activationAttribute == ActivationAttribute.PROVOKED;
    }

    public boolean isActive() {
        return activationAttribute == ActivationAttribute.ACTIVE;
    }


    public TargetRangeProperty getTargetRangeProperty() {
        if (setproperties == null)
            return null;
        for (Property prop : setproperties.getProperties()) {
            if (prop instanceof TargetRangeProperty)
                return (TargetRangeProperty) prop;
        }

        return null;
    }

    public TargetRelationProperty getTargetRelationProperty() {
        if (setproperties == null)
            return null;
        for (Property prop : setproperties.getProperties()) {
            if (prop instanceof TargetRelationProperty)
                return (TargetRelationProperty) prop;
        }

        return null;
    }

    /**
     * @param position
     * @return EffectTemplate
     */
    public EffectTemplate getEffectTemplate(int position) {
        return effects != null && effects.getEffects().size() >= position
                ? effects.getEffects().get(position - 1) : null;

    }

    /**
     * @return
     */
    public int getEffectsDuration() {
        return effects != null ? effects.getEffectsDuration() : 0;
    }

    /**
     * @return the cooldown
     */
    public int getCooldown() {
        return cooldown;
    }

    /**
     * @return the penaltySkillId
     */
    public int getPenaltySkillId() {
        return penaltySkillId;
    }

    /**
     * @return the pvpDamage
     */
    public int getPvpDamage() {
        return pvpDamage;
    }

    /**
     * @return the pvpDuration
     */
    public int getPvpDuration() {
        return pvpDuration;
    }

    /**
     * @return chainSkillProb
     */
    public int getChainSkillProb() {
        return chainSkillProb;
    }

    /**
     * @return cancelRate
     */
    public int getCancelRate() {
        return cancelRate;
    }

    public boolean isStance() {
        return stance;
    }

    /**
     * @return skillSetException
     */
    public int getSkillSetException() {
        return skillSetException;
    }

    public boolean hasResurrectEffect() {
        return getEffects() != null && getEffects().isResurrect();
    }

    public boolean hasItemHealFpEffect() {
        return getEffects() != null && getEffects().isItemHealFp();
	}
}

File: tools/DataTools/DataParsers/ParserBase/Skills/SkillTemplate.cs
namespace Jamie.Skills
{
	using System;
	using System.Collections.Generic;
	using System.Xml.Serialization;
	using System.ComponentModel;
	using System.Xml.Schema;
	using System.Reflection;
	using System.Diagnostics;
	using System.Linq;

	[Serializable]
	[XmlType(AnonymousType = true)]
	[XmlRoot("skill_data", Namespace = "", IsNullable = false)]
	public partial class SkillData
	{
		[XmlElement("skill_template", Form = XmlSchemaForm.Unqualified)]
		public List<SkillTemplate> SkillList = new List<SkillTemplate>();
	}

	[Serializable]
	[XmlType(AnonymousType = true)]
	public partial class SkillTemplate
	{
		[XmlElement(Form = XmlSchemaForm.Unqualified)]
		public Properties initproperties;

		[XmlElement(Form = XmlSchemaForm.Unqualified)]
		public Conditions startconditions;

		[XmlElement(Form = XmlSchemaForm.Unqualified)]
		public Properties setproperties;

		[XmlElement(Form = XmlSchemaForm.Unqualified)]
		public Conditions useconditions;

		[XmlElement(Form = XmlSchemaForm.Unqualified)]
		public Effects effects;

		[XmlElement(Form = XmlSchemaForm.Unqualified)]
		public Actions actions;

		[XmlAttribute]
		public int skill_id;

		[XmlAttribute]
		public string name;

		[XmlAttribute]
		[DefaultValue(0)]
		public int nameId;

		[XmlAttribute]
		public string stack;

		[XmlAttribute]
		public int lvl;

		[XmlAttribute]
		public skillType skilltype;

		[XmlAttribute]
		public skillSubType skillsubtype;

		[XmlAttribute]
		public TargetSlot tslot;

        [XmlAttribute]
        [DefaultValue(0)]
        public int tslot_level;

		[XmlAttribute]
		public activationAttribute activation;

		[XmlAttribute]
		public int cooldown;

		[XmlAttribute]
		public int duration;

		[XmlAttribute]
		[DefaultValue(0)]
		public int penalty_skill_id;

		[XmlAttribute]
		[DefaultValue(DispelCategory.NONE)]
		public DispelCategory dispel_category;

		[XmlAttribute]
		[DefaultValue(0)]
		public int dispel_level;

		[XmlAttribute]
		[DefaultValue(0)]
		public int pvp_damage;

		[XmlAttribute]
		[DefaultValue(0)]
		public int pvp_duration;

		[XmlAttribute]
		[DefaultValue(0)]
		public int cancel_rate;

		[XmlAttribute]
		[DefaultValue(0)]
		public int chain_skill_prob;

		[XmlAttribute]
		[DefaultValue(false)]
		public bool stance;

        [XmlAttribute]
        [DefaultValue(0)]
        public int delay_id;

		[XmlAttribute]
		[DefaultValue(0)]
		public int skillset_exception;

		public SkillTemplate() {
			this.cooldown = 0;
			this.chain_skill_prob = 0;
			this.cancel_rate = 0;
			this.stance = false;
			this.skillset_exception = 0;
		}

		public SkillTemplate(int skillId, int skillLevel) {
			this.skill_id = skillId;
			this.lvl = skillLevel;
		}
	}

	[Serializable]
	public enum skillSubType
	{
		NONE = 0,
		ATTACK = 1,
		CHANT = 2,
		HEAL = 3,
		BUFF = 4,
		DEBUFF = 5,
		SUMMON = 6,
		SUMMONHOMING = 7,
		SUMMONTRAP = 8,
	}

	[Serializable]
	public enum FlyRestriction
	{
		NONE = 0,
		ALL = 1,
		FLY = 2,
		GROUND = 3
	}

	[Serializable]
	public enum TargetSlot
	{
		NONE = 0,
		BUFF = 1,
		DEBUFF = 2,
		SPEC = 3,
		SPEC2 = 4,
		BOOST = 5,
		NOSHOW = 6,
		CHANT = 7,
	}

	[Serializable]
	public enum activationAttribute
	{
		NONE = 0,
		ACTIVE = 1,
		PROVOKED = 2,
		MAINTAIN = 3,
		TOGGLE = 4,
		PASSIVE = 5,
	}

	[Serializable]
	public enum DispelCategory
	{
		NONE = 0,
		ALL = 1,
		BUFF = 2,
		DEBUFF_MENTAL = 3,
		DEBUFF_PHYSICAL = 4,
		EXTRA = 5,
		NPC_BUFF = 6,
		NPC_DEBUFF_PHYSICAL = 7,
		STUN = 8,
	}
}
